compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala

```diff
@@ -510,18 +593,53 @@ abstract class GenJSCode extends plugins.PluginComponent
      *  The fields are initialized with the zero of their types.
      */
     def genClassFields(cd: ClassDef): List[js.FieldDef] = {
+      val classSym = cd.symbol
+      assert(currentClassSym.get == classSym,
+          "genClassFields called with a ClassDef other than the current one")
+
       // Non-method term members are fields, except for module members.
       (for {
-        f <- currentClassSym.info.decls
+        f <- classSym.info.decls
         if !f.isMethod && f.isTerm && !f.isModule
       } yield {
         implicit val pos = f.pos
+
         val mutable =
           suspectFieldMutable(f) || unexpectedMutatedFields.contains(f)
-        js.FieldDef(encodeFieldSym(f), toIRType(f.tpe), mutable)
+        val name =
+          if (isExposed(f)) js.StringLiteral(jsNameOf(f))
+          else encodeFieldSym(f)
+
+        val irTpe = {
+          if (f.tpe.typeSymbol != CharClass || !isScalaJSDefinedJSClass(classSym))
+            toIRType(f.tpe)
+          else
+            jstpe.ClassType(ir.Definitions.BoxedCharacterClass)
+        }
```

Comment:
> However, this doesn't handle value classes.

Indeed, [it doesn't](https://github.com/scala-js/scala-js/pull/1809/files#diff-6e8c360252085a4ec6f673609bd75631R238). I have not found a way to support this reliably so far. Fortunately this happens only for fields whose static type is a value class, and that are not initialized, i.e., they have `= _`. I guess we could forbid this in PrepJSInterop, now that I think about it.

> Also, why do we box private fields?

Consistency in the implementation. Private fields are rare enough and boxing cheap enough (basically no-op except for Char) that it is not worth complicating the implementation to specifically de-box private fields.